src/frontend/src/handler/alter_source_format_encode.rs

```diff
@@ -0,0 +1,309 @@
+            !col_a.is_hidden()
+                && !columns_b.iter().any(|col_b| {
+                    col_a.name() == col_b.name() && col_a.data_type() == col_b.data_type()
+                })
+        })
+        .cloned()
+        .collect()
+}
+
+pub async fn handle_alter_source_format_encode(
+    handler_args: HandlerArgs,
+    name: ObjectName,
+    connector_schema: ConnectorSchema,
+) -> Result<RwPgResponse> {
+    let session = handler_args.session;
+    let db_name = session.database();
+    let (schema_name, real_source_name) =
+        Binder::resolve_schema_qualified_name(db_name, name.clone())?;
+    let search_path = session.config().search_path();
+    let user_name = &session.auth_context().user_name;
+
+    let schema_path = SchemaPath::new(schema_name.as_deref(), &search_path, user_name);
+
+    let (database_id, schema_id, mut source) = {
+        let reader = session.env().catalog_reader().read_guard();
+        let (source, schema_name) =
+            reader.get_source_by_name(db_name, schema_path, &real_source_name)?;
+        let db = reader.get_database_by_name(db_name)?;
+        let schema = db.get_schema_by_name(schema_name).unwrap();
+
+        session.check_privilege_for_drop_alter(schema_name, &**source)?;
+
+        (db.id(), schema.id(), (**source).clone())
+    };
+
+    if source.associated_table_id.is_some() {
+        bail_not_implemented!("alter source associated with table is not supported yet");
+    }
+
+    let StreamSourceInfo {
+        format, row_encode, ..
+    } = source.info;
+    let (Some(old_format), Some(old_row_encode)) = (
+        format_type_to_format(FormatType::try_from(format).unwrap()),
+        encode_type_to_encode(EncodeType::try_from(row_encode).unwrap()),
+    ) else {
+        return Err(ErrorCode::NotSupported(
+            "altering a legacy source which is not created using `FORMAT .. ENCODE ..` Clause"
+                .to_string(),
+            "try this feature by creating a fresh source".to_string(),
+        )
+        .into());
+    };
+
+    if connector_schema.format != old_format || connector_schema.row_encode != old_row_encode {
+        bail_not_implemented!(
+            "the original definition is FORMAT {:?} ENCODE {:?}, and altering them is not supported yet.",
+            &old_format,
+            &old_row_encode,
+        );
+    }
+
+    if !schema_has_schema_registry(&connector_schema) {
+        return Err(ErrorCode::NotSupported(
+            "altering a source without schema registry".to_string(),
+            "try `ALTER SOURCE .. ADD COLUMNS ...`".to_string(),
+        )
+        .into());
+    }
+
+    let mut with_properties = source.with_properties.clone().into_iter().collect();
+    validate_compatibility(&connector_schema, &mut with_properties)?;
+
+    let create_cdc_source_job = if is_cdc_connector(&with_properties) {
+        CdcTableType::from_properties(&with_properties).can_backfill()
+    } else {
+        false
+    };
```

Comment:
cdc jobs cannot do the alter I think @StrikeW

Reply:
Yes. we should not allow user to alter the format and encoding of cdc source job. It is managed internally.